### PR TITLE
[inductor] Mask the 2D-tiled reduction tail store on the output axis

### LINKED ISSUE #196681: [inductor] CPU 2D-tiled reduction stores a full-width vector in the tail block: heap overflow (SIGABRT) or silent wrong results
### 🐛 Describe the bug

On CPU, Inductor generates a 2D-tiled kernel for a reduction whose non-reduction (tiled) dimension is not a multiple of the vector width. For the tail block, the reduction itself correctly uses `sum_masked_reduce(..., tail)`, but the final write-back is still a **full-width vector store**. This writes past the end of the output buffer (heap corruption / `SIGABRT`) and, when it does not crash, silently overwrites values in the next row.

```python
import torch
import torch.nn.functional as F

def f(x):
    v = (x * 0.5) * (torch.erf(x * 0.7071067811865476) + 1)   # GELU, x is channels_last
    a = v.permute(3, 2, 1, 0)                                  # (66, 66, 8, 1)
    p = F.pad(a, (0, 0, 0, 58), value=0.5)                     # (66, 66, 66, 1)
    q = F.pad(v, (0, 0, 0, 0, 0, 58), value=0.5)               # (1, 66, 66, 66)
    return torch.matmul(q, p)                                  # (66, 66, 66, 1)

torch.manual_seed(0)
x = torch.randn(1, 8, 66, 66).contiguous(memory_format=torch.channels_last)
ref = f(x)
out = torch.compile(f)(x)
print("max abs diff:", (out - ref).abs().max().item())
```

Default config, 3/3 runs with a fresh `TORCHINDUCTOR_CACHE_DIR`:

```
Fatal glibc error: malloc.c:2599 (sysmalloc): assertion failed: (old_top == initial_top (av) && old_size == 0) || ...
Aborted (core dumped)
```

Eager is fine. Using a contiguous input instead of `channels_last`, replacing the GELU with a single pointwise op (`erf`, `x * 2`, `relu`), or dropping the second pad all avoid it.

### Root cause (visible in the generated C++)

The `bmm` (N=1) is fused with the pads / GELU / permute into one kernel tiled 16x16 over `x0` (output dim, size 66) and `x2` (reduction dim, size 66), with a transposed gather into `tmpbuf` for `x0_inner`. For the `x0 = 64` tail (2 valid lanes) the accumulation is masked, but the store is not:

```cpp
// tail block over x0
tmp_acc0_vec = sum_masked_reduce(tmp_acc0_vec, tmp33, static_cast<int64_t>(2L));
...
tmp_acc0_vec.store(out_ptr0 + static_cast<int64_t>(x0 + 66L*x1));   // still a full 16-lane store when x0 == 64
```

The output buffer holds `4356 * 66 = 287496` floats. For the last row (`x1 = 4355`) this writes indices `287494..287509`, i.e. **14 floats past the end**. For every other row the 14 extra lanes land on elements `0..13` of the next row.

### Silent wrong results when it doesn't crash

With `MALLOC_MMAP_THRESHOLD_=65536` the output buffer is mmap'ed and the overflow falls into page slack, so the process survives and the values can be inspected (`OMP_NUM_THREADS=1`):

```
max abs diff 20.208 | wrong elems 60970/287496 | wrong cols [0..13] | first row wrong False
```

The wrong columns are exactly `x0 = 0..13` (the 14 spilled lanes), and the first row is never wrong because nothing spills into it. With 64 threads the set of corrupted rows varies between runs (it depends on whether the next row's first block is written before or after the spill).

### Relation to #181618 / #181691

This looks like the case #183699 (`[inductor] Fix tail reduction suffix width`, fixing #181618 and #181691) set out to fix — *"avoiding full-width vector stores into tail reduction buffers"* — but that change only applies to 1D tiling:

```python
self.aggregate_reduction_buffers(
    _inner_loop_reduction_outer_not
    or (len(tiling_indices) == 1 and _tiled_loop_reduction_descendant),
    _outer_loop,
)
```

Here the kernel is 2D-tiled (`len(tiling_indices) == 2`), so the condition is false and the full-width store remains. The fix commit is included in the nightly below, and the repro still crashes there.

### Versions

Reproduced on:
- `2.15.0.dev20260910+cpu` (nightly, git `682cb3eebadc`)
- `2.14.0+cpu`

CPU, Linux x86_64, Python 3.11, inductor backend.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by he-yufeng**
Fixes #196681

## Summary

A CPU 2D-tiled reduction whose two tiled axes are not adjacent loop levels (a pointwise level sits between them) stored the tail block's accumulator at full vector width. For a 66-wide output axis, the tail block at `x0=64` has 2 valid lanes but the epilogue emitted a full-width store, spilling into the next row on every row and past the buffer end on the last one — heap corruption or silent wrong results.

The store code is fine; the miss is one level up in the epilogue aggregation. `CppKernelProxy.aggregate_reduction_buffers` splits the reduction suffix into main/tail variants only when `_inner_loop_reduction_outer_not` holds, and that check probed `tiling_indices[0] + 1` for the inner reduction — assuming the second tiled axis is always the level right after the outer one. Here the tiled axes are levels 0 (output) and 2 (reduction) with a pointwise level in between, so the check read the wrong level, concluded "no inner reduction", and spliced only the main kernel's suffix. The tail kernels had already emitted the correctly masked store (`tmp_acc0_vec.store(ptr, 2)`); it never reached the output.

The fix probes the actual second tiled axis (`tiling_indices[1]` when two axes are tiled, the adjacent level otherwise as before). The epilogue then splits as designed:

```cpp
if(C10_LIKELY(x0 >= 0 && x0 < 64))   tmp_acc0_vec.store(out_ptr0 + x0 + 66*x1);
if(C10_UNLIKELY(x0 >= 64 && x0 < 66)) tmp_acc0_vec.store(out_ptr0 + x0 + 66*x1, 2);
```

## Test plan

New regression test in `test_cpu_repro.py` over the issue's channels_last GELU + pads + matmul shape (66 leaves a tail at every power-of-two vector width ≥ 2, so it exercises the tail on NEON and AVX alike):

```
python test/inductor/test_cpu_repro.py -k test_tile2d_reduction_masked_tail_store
```

Verified locally on CPU (Apple Silicon, torch 2.15.0.dev20260912 nightly): the test fails on main as shipped (3.0% of elements mismatched, max abs diff 19.5) and passes with this change; the same fix on the 2.14.0 release wheel takes the issue's repro from max abs diff 19.97 to 7.6e-06. A regression sweep over neighboring codegen shapes (2D-tiled pointwise, var_mean welford, argmax, mid-dim sum, and the 27-test reduction/tail/tile subset of test_cpu_repro) shows no outcome changes from the patch. The reduction/tail subset has 13 pre-existing failures on macOS (file-level `not IS_MACOS` gating means these run on Linux CI); the failure set is byte-identical with and without the patch. Ruff check/format on the changed lines is clean; full lintrunner was not run locally.

This PR was authored with AI assistance (Kimi); the mechanism analysis, the local verification described above, and the diff were reviewed by the committer before submission.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @laolvfan

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes
